src/data/moves/move-attrs/transform-attr.ts

```diff
@@ -27,3 +44,21 @@ export class TransformAttr extends MoveEffectAttr {
     return true;
   }
 }
+
+/**
+ * Helper function to check if a transform is valid
+ * @param user the Pokemon doing the transforming
+ * @param target the Pokemon that is the target of transform
+ * @returns whether or not the transform succeeds
+ */
+export function canTransform(user: Pokemon, target: Pokemon): boolean {
+  if (
+    user.hasTag(BattlerTagType.TRANSFORMED)
+    || target.hasTag(BattlerTagType.TRANSFORMED)
+    || target.hasTag(...SEMI_INVULNERABLE_BATTLER_TAG_TYPES)
+    || target.hasTag(BattlerTagType.SUBSTITUTE)
```

Comment:
This can be simplified by using the rest param that `hasTag` supports

```suggestion
    user.hasTag(BattlerTagType.TRANSFORMED)
    || target.hasTag(BattlerTagType.TRANSFORMED, ...SEMI_INVULNERABLE_BATTLER_TAG_TYPES, BattlerTagType.SUBSTITUTE)
```